File: health-platform/mobile/native/ios/VeepooChannelHandler.swift
import Foundation
import Flutter

// REFERENZ-IMPLEMENTIERUNG (Gerüst). Kompiliert erst mit eingebundenem
// Veepoo-iOS-Framework (VPBluetooth) und Flutter-Engine. Siehe ../README.md.
//
// Spiegelt den MethodChannel/EventChannel-Vertrag (veepoo_channel.dart) und
// mappt VPBleCentralManager-Callbacks in das einheitliche JSON-Protokoll
// (docs/07, docs/19). Werte NORMALISIERT, Zeit als Epoch-Millis.

// import VPBluetooth

final class VeepooChannelHandler: NSObject, FlutterStreamHandler {

    static let commands = "wearable/commands"
    static let events = "wearable/events"

    private var eventSink: FlutterEventSink?
    // private let manager = VPBleCentralManage.sharedBleManager()

    func register(with registrar: FlutterPluginRegistrar) {
        let methodChannel = FlutterMethodChannel(
            name: VeepooChannelHandler.commands,
            binaryMessenger: registrar.messenger())
        methodChannel.setMethodCallHandler(handle)

        let eventChannel = FlutterEventChannel(
            name: VeepooChannelHandler.events,
            binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(self)
    }

    // MARK: - MethodChannel

    private func handle(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let args = call.arguments as? [String: Any] ?? [:]
        switch call.method {
        case "scan":
            scan(timeoutMs: args["timeoutMs"] as? Int ?? 10000, result: result)
        case "connect":
            connect(id: args["id"] as! String,
                    password: args["password"] as? String ?? "0000",
                    result: result)
        case "disconnect": disconnect(result)
        case "capabilities": capabilities(result)
        case "deviceInfo": deviceInfo(result)
        case "battery": battery(result)
        case "sync": sync(sinceEpochMs: args["sinceEpochMs"] as? Int64, result: result)
        case "recordEcg": recordEcg(result)
        case "startRealtime": startRealtime(metric: args["metric"] as! String, result: result)
        case "stopRealtime": stopRealtime(result)
        default: result(FlutterMethodNotImplemented)
        }
    }

    // MARK: - EventChannel

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        eventSink = nil
        return nil
    }

    private func emit(_ type: String, _ payload: [String: Any?]) {
        var out = payload
        out["type"] = type
        eventSink?(out)
    }

    // MARK: - Kommandos (SDK-Aufrufe als TODO, Namen aus docs/19)

    private func scan(timeoutMs: Int, result: @escaping FlutterResult) {
        // TODO: manager.peripheralManage.veepooSDKStartScanDevice(...andReceiveScanningDevice:)
        //   -> [ { id, name, rssi } ]
        result([[String: Any]]())
    }

    private func connect(id: String, password: String, result: @escaping FlutterResult) {
        // TODO: veepooSDKConnectDevice:deviceConnectBlock: ; Passwort/Handshake im Connect.
        //   Capabilities aus VPPeripheralModel (deviceFuctionData, ecgType, hrvType, ...).
        //   connectionState -> emit("connectionState", ["state": "connected"])
        result(nil)
    }

    private func disconnect(_ result: @escaping FlutterResult) {
        // TODO: veepooSDKDisconnectDevice
        result(nil)
    }

    private func capabilities(_ result: @escaping FlutterResult) {
        // TODO: aus VPPeripheralModel-Flags interne Keys ableiten (nur real unterstützte).
        result([String]())
    }

    private func deviceInfo(_ result: @escaping FlutterResult) {
        result(["model": "Mecorly V500"])
    }

    private func battery(_ result: @escaping FlutterResult) {
        // TODO: veepooSDKReadDeviceBatteryInfo:
        result(nil)
    }

    private func sync(sinceEpochMs: Int64?, result: @escaping FlutterResult) {
        // TODO: veepooSdkStartReadDeviceAllDataWithReadStateChangeBlock:
        //   -> HR/SpO2/BP/HRV/Sleep/Steps sammeln -> ["measurements": [...]]
        result(["measurements": [[String: Any]]()])
    }

    private func recordEcg(_ result: @escaping FlutterResult) {
        // TODO nur wenn ecgType > 0: veepooSDKTestECGStart:testResult:
        result(nil)
    }

    private func startRealtime(metric: String, result: @escaping FlutterResult) {
        // TODO: heart_rate -> veepooSDKTestHeartStart:testResult: ; temperature -> veepooSDK_temperatureTestStart:result:
        //   emit("realtimeMeasurement", ["metric": metric, "value": ..., "unit": ..., "timeEpochMs": ...])
        result(nil)
    }

    private func stopRealtime(_ result: @escaping FlutterResult) {
        result(nil)
    }
}

File: health-platform/mobile/native/android/VeepooChannelHandler.kt
package app.vitaguide.wearable

// REFERENZ-IMPLEMENTIERUNG (Gerüst). Kompiliert erst mit eingebundener Veepoo-SDK
// (vpprotocol/vpbluetooth .aar) und Flutter-Engine. Siehe ../README.md.
//
// Übersetzt die MethodChannel/EventChannel-Kommandos (veepoo_channel.dart) in
// Aufrufe der VPOperateManager-API und mappt die SDK-Callbacks in das
// einheitliche JSON-Protokoll (docs/07, docs/19). Alle Werte NORMALISIERT,
// Zeitstempel in Epoch-Millis (UTC-Normalisierung erfolgt in Dart).

import android.content.Context
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

// import com.veepoo.protocol.VPOperateManager
// import com.veepoo.protocol.listener.data.*
// import com.veepoo.protocol.model.datas.*

class VeepooChannelHandler(
    private val context: Context,
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private var events: EventChannel.EventSink? = null
    // private val vp = VPOperateManager.getMangerInstance(context)

    companion object {
        const val COMMANDS = "wearable/commands"
        const val EVENTS = "wearable/events"
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "scan" -> scan(call.argument<Int>("timeoutMs") ?: 10000, result)
            "connect" -> connect(
                call.argument<String>("id")!!,
                call.argument<String>("password") ?: "0000",
                result,
            )
            "disconnect" -> disconnect(result)
            "capabilities" -> capabilities(result)
            "deviceInfo" -> deviceInfo(result)
            "battery" -> battery(result)
            "sync" -> sync(call.argument<Long>("sinceEpochMs"), result)
            "recordEcg" -> recordEcg(result)
            "startRealtime" -> startRealtime(call.argument<String>("metric")!!, result)
            "stopRealtime" -> stopRealtime(result)
            else -> result.notImplemented()
        }
    }

    // --- EventChannel ---
    override fun onListen(arguments: Any?, sink: EventChannel.EventSink?) {
        events = sink
    }

    override fun onCancel(arguments: Any?) {
        events = null
    }

    private fun emit(type: String, payload: Map<String, Any?>) {
        events?.success(payload + mapOf("type" to type))
    }

    // --- Kommandos (SDK-Aufrufe als TODO, Namen aus docs/19) ---

    private fun scan(timeoutMs: Int, result: MethodChannel.Result) {
        // TODO: vp.startScanDevice(SearchResponse{ onSearchResults -> Liste sammeln })
        // Ergebnis: List<Map> mit id (mac), name, rssi
        result.success(emptyList<Map<String, Any?>>())
    }

    private fun connect(id: String, password: String, result: MethodChannel.Result) {
        // TODO: vp.connectDevice(id, IConnectResponse, INotifyResponse)
        // Nach Verbindung: vp.confirmDevicePwd(...) mit
        //   IPwdDataListener (Firmware/Version -> PwdData)
        //   IDeviceFuctionDataListener (Capabilities -> DeviceFunctionPackage1..5)
        // connectionState-Events über emit("connectionState", {state: "connected"})
        result.success(null)
    }

    private fun disconnect(result: MethodChannel.Result) {
        // TODO: vp.disconnectWatch(IBleWriteResponse)
        result.success(null)
    }

    private fun capabilities(result: MethodChannel.Result) {
        // TODO: aus DeviceFunctionPackage1..5 die unterstützten Metriken ableiten
        //   und als List<String> mit den internen Keys (heart_rate, spo2, ...) liefern.
        //   NUR real gemeldete Fähigkeiten (docs/50) – z. B. ecg nur wenn ecgType > 0.
        result.success(emptyList<String>())
    }

    private fun deviceInfo(result: MethodChannel.Result) {
        // TODO: aus PwdData: deviceVersion (firmware), deviceNumber (model), serial
        result.success(mapOf("model" to "Mecorly V500"))
    }

    private fun battery(result: MethodChannel.Result) {
        // TODO: vp.readBattery(IBleWriteResponse, IBatteryDataListener -> BatteryData.batteryPercent)
        result.success(null)
    }

    private fun sync(sinceEpochMs: Long?, result: MethodChannel.Result) {
        // WICHTIG: SDK verträgt keine parallelen Operationen -> serielle Queue liegt
        // bereits in Dart (CommandQueue). Tages-Voll-Sync:
        // TODO: vp.readAllHealthDataBySettingOriginData(IAllHealthDataListener, day, position, watchday)
        //   -> HR (5-min), SpO2, BP (high/low), HRV, Sleep, Steps sammeln und in
        //      { "measurements": [ {metric,value,unit,timeEpochMs,ingestKey,quality} ] } mappen.
        result.success(mapOf("measurements" to emptyList<Map<String, Any?>>()))
    }

    private fun recordEcg(result: MethodChannel.Result) {
        // TODO nur wenn ecgType > 0: vp.startDetectECG(resp, isNeedCurve=true, IECGDetectListener)
        //   onEcgADCChange -> samples sammeln; Ergebnis { durationSeconds, sampleRate, samples[], timeEpochMs }
        result.success(null)
    }

    private fun startRealtime(metric: String, result: MethodChannel.Result) {
        // TODO metric-abhängig:
        //   heart_rate  -> vp.startDetectHeart(resp, IHeartDataListener.onDataChange)
        //   temperature -> vp.startDetectTempture(resp, ITemptureDetectDataListener)
        // Werte über emit("realtimeMeasurement", {metric, value, unit, timeEpochMs})
        result.success(null)
    }

    private fun stopRealtime(result: MethodChannel.Result) {
        // TODO: vp.stopDetectHeart(resp) / stopDetectTempture(...)
        result.success(null)
    }
}
